### ISSUE TITLE
p2 resolver does not follow redirects (e.g. http -> https)

### ISSUE THREAD

**Issue by reckart** [OpenNTF/p2-layout-provider#33]
Today, I noticed that P2 artifacts can no longer be resolved in one of my projects. The problem seems to be that nginx ships out HTML error page instead of an XHTML error page for artifacts that cannot be found:

`http://download.eclipse.org/releases/2018-12/compositeArtifacts.xml`
```
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

Note here that the `<hr>` tag is neither self-closing nor comes with a closing tag.

Btw. a similar page is also shipped by Maven Central `https://repo1.maven.org/maven2/org/eclipse/p2/201812/org.eclipse.jdt.debug.ui/maven-metadata.xml`. Maybe a general "update" in the default nginx error page.

```
<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
```

```
Downloading from apache.snapshots: https://repository.apache.org/snapshots/org/eclipse/p2/201812/org.eclipse.jdt.debug.ui/maven-metadata.xml
Downloading from central: https://repo1.maven.org/maven2/org/eclipse/p2/201812/org.eclipse.jdt.debug.ui/maven-metadata.xml
[Fatal Error] :6:3: The element type "hr" must be terminated by the matching end-tag "</hr>".
java.lang.RuntimeException: org.xml.sax.SAXParseException; lineNumber: 6; columnNumber: 3; The element type "hr" must be terminated by the matching end-tag "</hr>".
	at org.openntf.maven.p2.model.P2Repository.getBundles(P2Repository.java:95)
	at org.openntf.maven.p2.layout.P2RepositoryLayout.findBundles(P2RepositoryLayout.java:438)
	at org.openntf.maven.p2.layout.P2RepositoryLayout.lambda$9(P2RepositoryLayout.java:398)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1127)
	at org.openntf.maven.p2.layout.P2RepositoryLayout.getMetadata(P2RepositoryLayout.java:393)
	at org.openntf.maven.p2.layout.P2RepositoryLayout.getLocation(P2RepositoryLayout.java:156)
	at org.openntf.maven.p2.connector.P2RepositoryConnector.lambda$5(P2RepositoryConnector.java:120)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.xml.sax.SAXParseException; lineNumber: 6; columnNumber: 3; The element type "hr" must be terminated by the matching end-tag "</hr>".
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:257)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121)
	at org.openntf.maven.p2.util.xml.XMLDocument.loadInputStream(XMLDocument.java:65)
	at org.openntf.maven.p2.model.P2Repository.resolveCompositeChildren(P2Repository.java:147)
	at org.openntf.maven.p2.model.P2Repository.getBundles(P2Repository.java:77)
	... 10 more
```

Instead of trying to parse the error page, the resolver should follow the redirect:

`curl -I http://download.eclipse.org/releases/2018-12/compositeArtifacts.xml`
```
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Thu, 07 Oct 2021 07:20:49 GMT
Content-Type: text/html
Content-Length: 162
Connection: keep-alive
Location: https://download.eclipse.org/releases/2018-12/compositeArtifacts.xml
```

Observed with p2-layout-resolver v1.3.0 on Java 1.8.

**Comment by reckart** [comment]
Before trying to parse the error page as HTML, the resolver should look for status code and the `Location` header and follow it.

`curl -I http://download.eclipse.org/releases/2018-12/compositeArtifacts.xml`
```
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Thu, 07 Oct 2021 07:20:49 GMT
Content-Type: text/html
Content-Length: 162
Connection: keep-alive
Location: https://download.eclipse.org/releases/2018-12/compositeArtifacts.xml
```

**Comment by jesse-gallagher** [comment]
This should be fixed in 1.4.0. I added code to handle 302 as a redirect and to generally handle other non-redirect/non-OK statuses as failures. In my testing, that avoided the trouble seen here.
